core/jobs/batch_jobs/suggestion_stats_computation_jobs.py

```diff
@@ -97,98 +150,301 @@ def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
                 get_exploration_opportunity_summary_from_model)
             | 'Group by ID' >> beam.GroupBy(lambda m: m.id)
         )
+        skill_opportunities_by_id = (
+            self.pipeline
+            | 'Get all non-deleted skill opportunity models' >> (
+                ndb_io.GetModels(
+                    opportunity_models.SkillOpportunityModel.get_all(
+                        include_deleted=False)
+                )
+            )
+            | 'Transform to skill opportunity domain object' >> beam.Map(
+                opportunity_services.
+                get_skill_opportunity_from_model)
+            | 'Group skill opportunity by ID' >> beam.GroupBy(lambda m: m.id)
+        )
 
-        user_stats_results = (
+        exp_opportunity_to_submitted_suggestions = (
             {
-                'suggestion': suggestions_grouped_by_target,
+                'suggestion': translation_suggestions_grouped_by_target,
                 'opportunity': exp_opportunities
             }
             | 'Merge models' >> beam.CoGroupByKey()
-            | 'Get rid of key' >> beam.Values()  # pylint: disable=no-value-for-parameter
-            | 'Generate stats' >> beam.ParDo(
-                lambda x: self._generate_stats(
+            | 'Get rid of key submitted objects' >> beam.Values()  # pylint: disable=no-value-for-parameter
+        )
+        exp_opportunity_to_reviewed_suggestions = (
+            {
+                'suggestion': (
+                    reviewed_translation_suggestions_grouped_by_target),
+                'opportunity': exp_opportunities
+            }
+            | 'Merge reviewed models' >> beam.CoGroupByKey()
+            | 'Get rid of key of reviewed objects' >> beam.Values()  # pylint: disable=no-value-for-parameter
+        )
+        skill_opportunity_to_submitted_suggestions = (
+            {
+                'suggestion': question_suggestions_grouped_by_target,
+                'opportunity': skill_opportunities_by_id
+            }
+            | 'Merge submitted question models' >> beam.CoGroupByKey()
+            | 'Get rid of key of submitted question objects' >> beam.Values()  # pylint: disable=no-value-for-parameter
+        )
+        skill_opportunity_to_reviewed_suggestions = (
+            {
+                'suggestion': reviewed_question_suggestions_grouped_by_target,
+                'opportunity': skill_opportunities_by_id
+            }
+            | 'Merge reviewed question models' >> beam.CoGroupByKey()
+            | 'Get rid of key of reviewed question objects' >> beam.Values()  # pylint: disable=no-value-for-parameter
+        )
+
+        translation_contribution_stats_keys_and_results = (
+            exp_opportunity_to_submitted_suggestions
+            | 'Generate translation contribution stats' >> beam.ParDo(
+                lambda x: self._generate_translation_stats(
```

Comment:
Hi @seanlip I am trying to wrap up this PR but still stuck in the case when topic_id is None case.

Here we call _generate_translation_stats method with the suggestion and corresponding opportunity. But if the opportunity is None, we just pass None. https://github.com/chris7716/oppia/blob/d83d644ec0cea381d7b99b4056526c7d349bd752/core/jobs/batch_jobs/suggestion_stats_computation_jobs.py#L430

In that case topic_id would be an empty string when computing stats.

How should I handle this case? Should I just omit stats if the opportunity is not available?

Reply:
Hi @chris7716 -- could you please describe in what cases topic_id is None? What situations result in that happening?

Reply:
Sorry I have incorrectly mentioned as **topic_id None case**. It should be corrected as opportunity of a given suggestion is None case.

Let's say there is a translation and the corresponding opportunity is deleted after the translation was made.

According to this https://docs.google.com/document/d/1wM9cQzq1-3nbEhZliRlpnGDXbM_HspNkY16CYnA6lWg/edit#bookmark=id.gsi25q7oi77h opportunities can be deleted when:

- Their corresponding explorations are deleted
- Their corresponding stories are deleted or unpublished
- Their corresponding topics are deleted or unpublished

So in that case opportunity of the corresponding suggestion would be None and then the topic_id for stats will be and empty string according to this implementation https://github.com/chris7716/oppia/blob/d83d644ec0cea381d7b99b4056526c7d349bd752/core/jobs/batch_jobs/suggestion_stats_computation_jobs.py#L430

Reply:
@chris7716 and I discussed this offline and decided to omit stats with no topic opportunities, since we do not have a proper plan for displaying those stats to translators. In the future, if we change this policy, we can regenerate the stats.

@chris7716 also mentioned that he will create a followup issue to update translation stats when removing lessons.